src/theory/ff/sub_theory.h

```diff
@@ -0,0 +1,168 @@
+/******************************************************************************
+ * Top contributors (to current version):
+ *   Alex Ozdemir
+ *
+ * This file is part of the cvc5 project.
+ *
+ * Copyright (c) 2009-2022 by the authors listed in the file AUTHORS
+ * in the top-level source directory and their institutional affiliations.
+ * All rights reserved.  See the file COPYING in the top-level source
+ * directory for licensing information.
+ * ****************************************************************************
+ *
+ * A field-specific theory
+ */
+
+#include "cvc5_private.h"
+
+#ifdef CVC5_USE_COCOA
+
+#ifndef CVC5__THEORY__FF__SUB_THEORY_H
+#define CVC5__THEORY__FF__SUB_THEORY_H
+
+#include <CoCoA/RingFp.H>
+
+#include <optional>
+#include <string>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+#include "context/cdlist_forward.h"
+#include "context/cdo.h"
+#include "context/context.h"
+#include "expr/node.h"
+#include "options/ff_options.h"
+#include "smt/env_obj.h"
+#include "theory/ff/core.h"
+#include "theory/ff/groebner.h"
+#include "theory/ff/stats.h"
+#include "theory/theory.h"
+#include "util/integer.h"
+
+namespace cvc5::internal {
+namespace theory {
+namespace ff {
+
+/**
+ * A solver for a specific, known finite field.
+ *
+ * While the main ff solver is responsible for all elements in any finite field,
+ * this solver just considers a single field. The main ff solver essentially
+ * just multiplexes between different sub-solvers.
+ *
+ * The solver only exposes a subset of the standard SMT theory interface. See
+ * the methods below.
+ *
+ * For now, our implementation assumes that the finite field has prime order.
+ */
+class SubTheory : protected EnvObj, protected context::ContextNotifyObj
+{
+ public:
+  // Create a new sub-theory.
+  //
+  // Parameters:
+  // * modulus: the size of this field for this theory, a prime.
+  SubTheory(Env& env, FfStatistics* stats, Integer modulus);
+
+  // Notify this theory of a node it may need to handle.
+  //
+  // All nodes this theory sees in the future must be pre-registered.
+  void preRegisterTerm(TNode);
+
+  // Assert a fact to this theory.
+  void notifyFact(TNode fact);
+
+  // Check the current facts.
+  void postCheck(Theory::Effort);
+
+  // Has a conflict been detected?
+  bool inConflict() const;
+
+  // What is that conflict?
+  const std::vector<Node>& conflict() const;
+
+  // Get a model.
```

Comment:
I assume this can only be called at a certain point? E.g. after full effort check? If so, document this.